readme.txt
*** Link 68 (link68) version 0.2 (alpha) README ***
Program and documentation copyright (c) Joseph Herning 2021.
This is experimental software. Please read license at the end of this file.

Link 68 is a pure Python computer-to-calc linking program for the TI 68k calculators. It does not make use of external libraries except those Python modules needed to support various things (PySerial, PyUSB, PyPNG...). As of release, testing has only been done on Linux. This manual assumes a Linux-like environment. Current features/capabilities:
 - sending/receiving normal variables using TI Graph Link format (including group format)
 - backing up all normal variables (i.e. not flash apps or AMS)
 - sending free flash apps and AMS updates
 - screenshots
 - listing calculator contents and simple variable searches
 - grouping and ungrouping variable files (TI Graph Link format, not .tig)

BASIC USAGE (assuming sh-type shell):
    ./link68.py COMMAND ARGS [optional switches]

 - Optional switches do not need to be at the end.

 
 --- Setup ---

Before running, set your link type and necessary port settings in settings.py. See the Link Cable Notes section. You may wish to look through the other settings as well. You will need a recent version of Python 3 (tested on 3.8). For serial cables, you will need PySerial. For USB cables you will need PyUSB. For screenshots, you will need PyPNG.

Make sure calculator is on and at the home screen before running, or you may have link issues. All transfers are silent and do not require additional action on the calculator.

You may need to make link68.py executable:
    chmod +x link68.py
If this doesn't work, you may need to modify the environment string at the beginning of the file or invoke your Python interpreter instead of using ./link68.py to run the program:
    <Python interpreter invocation> link68.py COMMAND ARGS [optional switches]
    
A good idea might be to create a new folder for your calculator data inside the source folder, e.g. link68/calcstuff. Then, working from the calcstuff folder, simply invoke link68 as "../link68.py" instead of "./link68.py". This will keep the program source files and your data from getting mixed together.

    
 --- Linking Examples ---

Take screenshot, save to myshot.png:
    ./link68.py shot myshot
 - Writes screenshot.png if no filename given.
 - Link read speed is also reported. PyPNG is required.

List all variables/apps on calculator:
    ./link68.py ls

Find variables on calculator matching a search string:
    ./link68.py find myfold/
 - This will list all the vars in myfold. A variable named "myfold" won't match because of the "/".
 - If multiple strings are given, vars matching ANY string are reported.

Get myvar1, myvar2, and myvar3 from calculator current directory:
    ./link68.py get myvar1 myvar2 myvar3
 - Files will be saved as main.myvarX.EXT where EXT is automatic.
 - The calculator does not report the current folder, so "main" is used.
 - get/getg are for normal vars; getting flash apps is not supported.

Get main/myvar1 and myfold/mayvr4:
    ./link68.py get main/myvar1 myfold/myvar4
 - You will get main.myvar1.EXT and myfold.myvar4.EXT
 - Computer files can be safely renamed. Metadata stores the on-calc var name and folder.

Get some files and save to a group file:
    ./link68.py getg main/myvar1 myfold/myvar4 groupfilebasename
 - As with normal files, the file extension is automatic (89g, 9xg, 92g).
    
Find variables then automatically get them: [Combines find and get/getg.]
    ./link68.py findget myfold/
    ./link68.py findgetg myfold/ groupfile

Put some files to the calculator (normal vars and/or group files):
    ./link68.py put myfold.nice.89p myfold2.fun.89e

Put some files to the current folder (disregarding folder metadata):
    ./link68.py put myfold.nice.89p myfold2.fun.89e -co
 - ** -co implicitly enables overwrite mode. **

Backup all normal vars to a groupfile:
    ./link68.py backup backupfile
 - If no file is provided, backup.XXg is used.

Flash the calculator with a flash apps or AMS:
    ./link68.py flash flashfile1.XXk/XXu flashfile2.XXk ...
 - AMS flashing may brick your calculator and is untested with many hardware/AMS combinations. See below.
 
 "Recover"/flash the calculator from the bootloader (after pressing "I to install product code"):
     ./link68.py rflash osfile.XXu
 - Bootloader gets confused by link68's normal calculator autodetection, so we need a special command.
 - Note: Bootloader is a little picky (especially on the TI-92+), so this MAY NOT work even if normal OS flashing is working. I have tested it on at least one 89/92+/V200 though.
 

 --- Optional Switches ---
 
-o  (Overwrite):
    link68 performs a check before overwriting files on the computer or vars on the calc (requiring a directory listing from the calc). The default behavior is to skip potential overwrites, but this forces overwriting.
 
-co  (Current folder + Overwrite):
    link68.py puts vars to current folder on calc instead of to the folder in the file metadata. Also enables overwrite mode since we do not know the current folder on the calc to perform overwrite checks.

-89 / -92 / -92+
 - Change file writing format when using group/ungroup. Only applies to group/ungroup. Examples below.


 --- Utility Commands (link unused) ---

For group and ungroup commands, all variables are loaded from the file(s) specified into link68's internal representation then file(s) are written from that. Vars can be loaded from any 68k var file or group file (regardless of calculator type). 'Group' always writes a single groupfile while 'ungroup' always writes individual files. The calculator type metadata / file extensions are determined by 'calc_type' in settings.py. You can override this with -92, -92+, or -89. This writes the appropriate extension and changes the file metadata but does not alter the raw variable data.

Create a groupfile from multiple files. The groupfile base name is a required final argument:
    ./link68.py group file1.ext file2.ext groupfile

Ungroup a groupfile to multiple files:
    ./link68.py ungroup groupfile.89g

Convert a TI-92 expression file to a TI-89 one:
    ./link68.py ungroup exp.92e -89
Convert several at once:
    ./link68.py ungroup exp.92e prog.92p mat.92m -89
Convert a whole bunch:   [Requires BASH-like shell.]
    ./link68.py ungroup *.92* -89
Instead, convert to a groupfile:
    ./link68.py group exp.92e prog.92p mat.92m groupfile -89


 --- Link Cable Notes ---
 
For all cables: the current userspace Python "drivers" are quick-and-dirty. I really just wanted to get all official TI cables for the classic link port working to some degree.

** Gray TI Graphlink (Serial) **
This cable is, in my experience, the most compatible. (TI Graph Link can use it on Linux through WINE!) PySerial is required. To connect to a modern computer you will probably use a USB-to-RS232 converter. Many cheap converters do not support all modem control lines of the serial port, and the cable draws power from two of them. Most USB-to-RS232 ICs are capable of controlling the lines, but the lines are often not implemented either in the hardware or in the driver. I have had luck with several converters using pl2303 and ft232r chips. The link rate is about 1 kilobyte/s (9600 baud without flow control!). Currently the ~2s link timeout is not implemented.

Your user will need access to the serial ports. This is often means the user should be in the "dialout" group.

** Custom Serial Link **
This is identical to the Gray Link driver except it does not switch-on the modem control lines which power the Gray Link. Use it to support your experimental/homebrew serial link (Note the $4 "serial" link is not actually a serial link -- try the Black Link driver). Modify the serial port settings (baud rate, flow control, etc.) in customlink.py as needed. It could also be used as a template for implementing a new non-serial link.

** Silver TI Graphlink (USB) **
This cable is supported through PyUSB which typically uses libUSB (but can have various backends). This cable seems a little more finicky and requires some USB pipe resetting between transfers to keep from choking. Warning: I know very little about USB/USB programming. Still, it works well for me at the moment and link rate is about 5-6 kilobytes/s as expected. There are several revisions, some using TI microcontrollers and some older ones using Cypress chips. The Cypress cables seem about 1 kilobyte/s faster than the TI ones.

The cable is found on the USB bus by its vendor and product IDs, so only have one plugged at a time. You may need to grant permission for user access to the USB device on your system. On Ubuntu-ish systems the following may work:

Create a udev rule, something like "/etc/udev/rules.d/51-silverlink.rules" consisting of the line:
    SUBSYSTEM=="usb", ATTR{idVendor}=="0451", ATTR{idProduct}=="e001", MODE="0660", GROUP="plugdev", SYMLINK+="silverGL%n"
Then reload the udev rules:
    sudo udevadm control --reload-rules
Make sure your user is in the "plugdev" group. Re-plug and cross your fingers.
I have no idea what you might need to do on Windows.

** Black "Windows" TI GraphLink **
I wanted to get this cable working out of nostalgia as I had one "back in the day." PySerial is required. This cable connects to an RS232 serial port but is not an actual asynchronous serial device like the gray cable. The computer bit-bangs the modem control lines, toggling/reading several times per bit. The code is untested on a real UART (e.g. 16550) and is a bit hackish due to the limitations of PyUSB (modem control lines cannot be set at the same time except if the port is closed). I've tried several USB-to-RS232 adapters but have only had luck with FTDI ft2332r ones, and the speed is around 300 bytes/s. This is due to USB latency since several USB operations must happen per bit. Also remember that some ft232r adapters may not implement the necessary modem control lines in hardware. Homemade "$4 serial link" cables may work but are untested.

Your user will need access to the serial ports. This is often means the user should be in the "dialout" group.
 
 
  --- Other Notes, Some Being Important ---
  
 - Forward (/) and back (\) slashes can both be used between a folder and variable name on the calculator, but in most Linux shells "\" must be escaped. e.g.: "main\\myvar". I think it's easier just to use a forward slash: "main/myvar". I don't know how Windows behaves.

 - TI-89 Titanium is untested. It should work through its classic link port at least for regular (non-flash) vars. Future plans INCLUDE testing it and supporting its direct USB link port.
 
 - HW1 calcs (including the TI-92 with Plus Module) are currently untested.
 
 - TI-89 (classic), TI-92, TI-92+, and V200 have all been tested, but not all physical-link/command/calc combinations have been tested. Most old AMS versions have not been tested.

 - When writing to computer, file extensions are added automatically; don't include them in filenames. When reading files from computer, full filenames should be used.
 
  - On most Linux-like environments you can use commands like "./link68.py put *" since "*" is expanded by the shell.

 - No calculator-type checking is performed when sending vars. That is, link68 is happy to send prog.92p to a TI-89. No adjustment is made to the variable data, but 68k calc variables are mostly compatible between calcs anyway, with exceptions including flash apps and ASM progs. Flash apps sent to the wrong calculator type will probably be rejected by the calculator at the end of transfer. This will essentially delete the existing version of the flash app from the calc.

 - Locked vars are problematic when in overwrite mode. Attempting to overwrite a locked var on the TI-92 seems to lead to a link timeout. Other calcs complain more gracefully.
 
  - Since the TI uses a modified ASCII character set to support Greek letters, etc., we currently convert these to their character codes in hex. For example, the mathtools package contains an expression named with two gammas on the calc. Link 68 will call it "+83+83" (83h being the code for a gamma in TI's charset). You can "./link68.py get mathtool/+83+83", and you will get the file "mathtool.+83+83.89e". If you send it back to the calculator it will have its original name. Var names and folders are stored in metadata and are not dependent on filenames, so feel free to rename your computer files (e.g. to remove the folder names).
  
   - Variable attributes (locked/unlocked, archived) are written to the computer files in what should be a TiLP compatible format (untested). Attributes are NOT restored when putting files to calculator which can lead to memory problems when restoring backups. This is the behavior of TI Graph Link since the link protocol for sending a var to the calc does not specify attributes. Programs seem to be an EXCEPTION. Locked programs are restored as locked which implies this information is stored in the raw variable data. The Link Guide does not seem to have information about this. Potential future support for remote control commands would allow us to change attributes after sending.
 
 
  --- Future Plans ---
  
  - Test on the TI-89 Titanium and support the direct USB cable.
  - Test on some older AMS versions.
  - Polish code.
  - Better error checking & exception handling.
  - Support receiving flash files -- MAYBE.
  - Support remote commands -- MAYBE. Could be used to implement variable deletion, empty folder creation/deletion, archiving files when restoring a backup.
  
  
  --- Things I DON'T plan to do ---
  
   - Support the Z80 calculators or anything besides the 68k calcs.
   - Add a GUI.
   - Support for certificates or non-free apps.
   - Support CBL/CBR.


 --- WHY? ---
 
 Motivations for writing Link 68 include:
  - Programming practice: I haven't programmed anything complicated (besides LaTeX) in quite some time, so link68 was a chance to brush up on skills and learn more about Python.
  - I like using the 68k calcs and want to continue to use them in the future.
  - Historical preservation: The 68k calcs are end-of-life. TI will be phasing out the TI-89 Titanium if they're not already, and the TI-92 was over 25 years old when I started this project! Open source projects for supporting vintage platforms are critical to preserving computing history.
  - No fullish-featured, CLI linking program exists supporting all the 68k calcs.
  - I wanted a self-contained program with minimal dependencies.
  - I want to tinker with homebrew link cables.
  
  
 --- Thank You! ---
 
  - Tim Singer and Romain Liévin, authors of the Link Protocol Guide.
  - Other members of the TI-calculator hacking community who have posted useful information on various websites and message boards over the years.
  - TI for making an interesting calculating platform, regardless of their frequent ambivalence/animosity to the user and hacker communities.
 
 
 --- LICENSE ---
 
 Copyright (c) 2021 Joseph Herning

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.



changelog.txt
2021-08-04: Added link speed reporting for get, put, and related commands.
2021-08-02: Fixed flash file parser to skip over license headers. Mimics TI Connect 1.2.1 in this regard, but does warn user that a license is embedded in the file.
2021-08-02: Fixed issue with screenshots when calc reports not ready.


link68.py
#!/usr/bin/env python3

from sys import argv
from os.path import exists
import time
# Now get Joe's modules:
from prot import *
from tifiles import *
import settings
from univ import *

# Used LOUD as its own command and not LOUD for find / overwrite checks
# Gets a listing from the calc and returns as a calcdir object.
def ls(link, loud):
    if loud: print('Getting directory listing..')
    entries = []
# Should get folders, files, and flash apps
    sendglobalREQ(link)
    quickget('ACK', link)
    quickget('VAR', link) # disregarding
    # Loop until EOT (needed for 92 only):
    calcdata = calcdir()
    while True:
        quicksend('ACK', link)
        quicksend('CTS', link)
        quickget('ACK', link)
        DATApack = quickget('DATA', link)
        calcdata.parse_dirDATA(DATApack, loud)
        quicksend('ACK', link)
        if getpacket(link).CIDt == 'EOT': break
    quicksend('ACK', link)
    return(calcdata)

# List matching var path/names, return list of matching path/names.
def find(args, link, loud):
    if loud: print('Getting directory listing to search..')
    calclist = ls(link, False)
    filelist = []
    for strtofind in args:
        strtofind = strtofind.replace('/', '\\')
        for entry in calclist.files:
            if entry.find(strtofind) != -1:
                entry = entry.replace('\\', '/')
                filelist.append(entry)
                if loud: print('  ' + entry)
    return(filelist)

# Combine find and get.
def findget(args, link, grouped):
    if len(args) < 2:
        print('Usage: findgetg [string(s) to match] group_file_base_name')
        quit()
    if grouped: groupfile = args.pop() # Remove group file name before searching.
    filelist = find(args, link, False)
    if grouped: filelist.append(groupfile)
    get(filelist, link, grouped)

# Backup all variables to a group file.
def backup(args, link):
    if len(args) == 0: backupname = 'backup'
    else: backupname = args[0]
    print('Getting directory listing..')
    calclist = ls(link, False)
    getargs = calclist.files
    print('  ' + str(len(getargs)) + ' vars to backup.')
    getargs.append(backupname) # Get needs the group file base name.
    get(getargs, link, True)

# Load TI vars from computer files. Return list of tivar objects.
def loadvars(filenames):
    tivars = []
    for filename in filenames:
            tivars.extend(parse_tifile(filename))  # Files may have multiple vars.
    return(tivars)

# Load TI vars from computer then write individually.
def ungroup(filenames):
    tivars = loadvars(filenames)
    for var in tivars:
        writevars(var.varname, [var], settings.calc_type)
# ungroup 'folder.something.89p -92' Will write folder.something.9xp

# Load TI vars from computer then write as a group file.
def group(filenames):
    if len(filenames) < 3:
        print('Usage: group [files to group] group_file_base_name')
        quit()
    pcfile = filenames.pop()
    tivars = loadvars(filenames)
    writevars(pcfile, tivars, settings.calc_type)

# Write flash file(s) to calc.
def flash(filenames, link, recovery):
    flashvarsALL = []
    for ffile in filenames: # Load flash files.
        flashvarsALL.extend(parse_flashfile(ffile))
    
    flashvars = [] # Check for overwrites
    if (not settings.overwrite) and (not recovery):
        print("Getting directory listing for overwrite check..")
        calclist = ls(link, False)
        for var in flashvarsALL:
            if var.name in calclist.apps:
                print('** ' + var.name + ' app exists. Skipping. **')
            else:
                flashvars.append(var)
    else: flashvars = flashvarsALL
    
    if recovery: # In recovery mode, we haven't auto-detected anything.
        settings.comp_MIDt = 'Computer->89/92+/V200' # No TI-92s in this rodeo.
    
    for var in flashvars: # Flash the Calc!
        link.softreset()  # Good for SilverLink
        if var.TID == 0x23 and var.hw_id >= 8: # AMS flash V200/89T!
            sendV200amsRTS(len(var.data), var.hw_id, link) # 0x08: V200, 0x09: Titanium
        elif var.TID == 0x23: # AMS flash 89/92+
            send8992amsRTS(var.name, len(var.data), link)
        else:
            sendRTS('', var.name, var.TID, len(var.data), link) # App
        ## Send 64k at a time:
        while (var.data):
            quickget('ACK', link)
            quickget('CTS', link)
            quicksend('ACK', link)
            outpack = packet()
            outpack.data.extend(var.data[0:64*1024])
            var.data = var.data[64*1024:] # slice off data we'll be sending (kills object)
            outpack.CIDt = 'DATA'
            print(' Flashing ' + var.name + '  (' + str(len(outpack.data)) + ' byte block).')
            outpack.buildsend(link)
            quickget('ACK', link)
            if var.data: # There's more!
                quicksend('CONT', link)
        quicksend('EOT', link)
        quickget('ACK', link)
    print('Done flashing.')
    if recovery: quit()

# Load then put normal vars to calculator.
def put(filenames, link):
    data_time = 0
    data_amount = 0
    putvarsALL = loadvars(filenames)
    putvars = []
    if not settings.overwrite:
        print("Getting directory listing for overwrite check..")
        calclist = ls(link, False)
        for var in putvarsALL:
            if var.folder + '\\' + var.varname in calclist.files:
                print('** ' + var.folder + '/' + var.varname + ' exists. Skipping. [Overwrite: -o] **')
            else:
                putvars.append(var)
    else: putvars = putvarsALL
            
    for var in putvars:
        link.softreset()  # SilverLink..
        if settings.current: # Send to current dir.
            var.folder = ''
        sendRTS(var.folder, var.varname, TID_ft[var.TIDt], len(var.vardata), link)
        quickget('ACK', link)
        quickget('CTS', link)
        quicksend('ACK', link)
        # Build and send the DATA packet:
        outpack = packet()
        outpack.data.extend(b'\x00\x00\x00\x00')
        outpack.data.extend(var.vardata)
        outpack.CIDt = 'DATA'
        print(' put ' + var.folder + '/' + var.varname + '  (' + var.TIDt + ', ' + str(len(var.vardata)) + ' bytes)' )
        timer = time.time()
        outpack.buildsend(link)
        data_time += time.time() - timer
        data_amount += len(outpack.output)
        ### /DATA
        quickget('ACK', link)
        quicksend('EOT', link) # Send EOT for each file. FIX?
        quickget('ACK', link) # Doesn't seem to work w/o doing this.
    if data_time: print("Done putting files. Link write speed: %i bytes/s" % (data_amount/data_time) )

# Get vars from calc; write individually or as a group file.
def get(args, link, grouped):
    data_time = 0
    data_amount = 0
    sendvars = []
    if grouped:
        groupfilename = args.pop()
    for filename in args:
        print("get " + filename)
        filename = filename.replace("/", "\\")  # Rewrite forward slash from command line.
        sendvarREQ(filename, link)
        quickget('ACK', link)
        headerpack = getpacket(link) # Is it a VAR or SKE?
        if headerpack.CIDt == 'VAR':
            quicksend('ACK', link)
            quicksend('CTS', link)
            quickget('ACK', link)
            timer = time.time()
            datapack = quickget('DATA', link)
            data_time += time.time() - timer
            data_amount += datapack.data_len + 6
            quicksend('ACK', link)
            quickget('EOT', link)
            quicksend('ACK', link)
            # Now parse the header/data.
            calcvar = tivar()
            calcvar.parse_silentreq(headerpack, datapack)
            # If invoked as "folder\filename", set folder.
            if "\\" in filename:
                calcvar.folder = filename.split("\\")[0]
            else:
                calcvar.folder = 'main' # If no dir specified, set to main. FIX?
            if grouped:
                sendvars.append(calcvar) # Add to list for group send later.
            else:
                writevars(filename, [calcvar], settings.calc_type)
        elif headerpack.CIDt == 'SKE':
            print(' ** Got SKE from calc. ' + filename + ' probably does not exist **')
            time.sleep(2.1) # Calculator seems to timeout after SKE.
                            # Then graylink gets a byte, so reset:
        link.softreset() # SilverLink!
    if data_time: print("Done getting files. Link read speed: %i bytes/s" % (data_amount/data_time) )
    if grouped:
        writevars(groupfilename, sendvars, settings.calc_type)

def screenshot(args, link):
    import png
    print('Getting screenshot.')
    quicksend('SCR', link)
    quickget('ACK', link)
    timer = time.time()
    inpack = quickget('DATA', link)
    timer = time.time() - timer
    print ('  Link rate: %i bytes/s' % ((inpack.data_len + 2)/timer))
    quicksend('ACK', link)

    sshot = [[0 for i in range(240)] for j in range(128)]  # 240 cols, 128 rows
    for row in range(0, 128):
        for colbyte in range(0,30): 
            recvd = inpack.data[row*30 + colbyte]
            for i in range(8):
                sshot[row][colbyte*8+7-i] = int(recvd&(2**i) == 0)

    sshot89 = [[0 for i in range(160)] for j in range(100)]  # 160x100
    for row in range(0,100): # Crop for TI-89
        for col in range(0,160):
            sshot89[row][col] = sshot[row][col]
    if len(args) > 0:
        filename = args[0] + '.png'
    else: filename = 'screenshot.png'
    if exists(filename) and settings.overwrite == False:
        print('** ' + filename + ' exists. Quitting. [Overwrite: -o] **')
        quit()
    with open(filename, 'wb') as f:
        if settings.calc_type == 'TI-89':
            w = png.Writer(160, 100, greyscale=1, bitdepth=1)
            w.write(f, sshot89)
        else:
            w = png.Writer(240, 128, greyscale=1, bitdepth=1)
            w.write(f, sshot)
        f.close()
        print('  Wrote ' + filename + '.')

#### MAIN ####

# Parse argv[] into cmd and args[], process flags.
argstemp = argv
argstemp.pop(0)
args = []
for el in argstemp: # Process any flags:
    if el == '-o':
        settings.overwrite = True
    elif el == '-co':
        settings.current = True
    elif el == '-89':
        settings.calc_type = 'TI-89'
    elif el == '-92':
        settings.calc_type = 'TI-92'
    elif el == '-92+':
        settings.calc_type = 'TI-92+'
    else: args.append(el)
    
if settings.current: settings.overwrite = True

if settings.current: print('** Current folder mode enabled. **')
if settings.overwrite: print('** Overwrite mode enabled. **')    

if len(args) == 0:
    print('Link 68 version 0.2 (alpha)')
    print('Usage: link68.py COMMAND ARGS [optional switches]')
    print(' COMMANDS: ls, shot, get, getg, put, find, finget, findgetg, flash, rflash, backup')
    print(' OPTIONS: -o: overwrite mode, -co: current folder + overwrite mode')
    print(' Settings contained in settings.py.')
    quit()
if args: cmd = args.pop(0)

    
# Do any commands that don't require the link cable, then quit:
if cmd == 'ungroup':
    ungroup(args)
    quit()
if cmd == 'group':
    group(args)
    quit()

# Init the link cable.
if settings.link_type == 'gray':
    from graylink import glink
elif settings.link_type == 'black':
    from blacklink import glink
elif settings.link_type == 'silver':
    from silverlink import glink
elif settings.link_type == 'custom':
    from customlink import glink
link = glink()

# Skip connectivity test for rflash:
if cmd == 'rflash': flash(args, link, True)

# Check connectivity & detect calculator type
settings.comp_MIDt = 'Computer->92' # Pretend we think it's a 92. Later models respond (as a 92), so it's safe.
quicksend('RDY', link)
inpack = quickget('ACK', link)
if inpack.header[3] & 1 == 1 and cmd != 'shot':
    print("Calc not ready. Are you at the homescreen?")
    quit()
if inpack.header[2] == 0x00: # Byte 3 of ACK seems to contain hardware version info.
    # Known responses: 0x0c: TI-92+ w/AMS 2.09, 0x04: TI-92+ w/ AMS 2.05
    # We are assuming all plain 92's give 00. TI-92-II???
    settings.calc_type = 'TI-92'
else:
    settings.comp_MIDt = 'Computer->89/92+/V200' # It's not a 92. Should ACK as 89 or 92+/V200.
    quicksend('RDY', link)
    inpack = quickget('ACK', link)
    settings.calc_type = inpack.MIDt
print('  ' + settings.calc_type  + ' detected.')

# Do requested linking command.
if cmd == 'shot': screenshot(args, link)
elif cmd == 'get': get(args, link, False)
elif cmd == 'getg': get(args, link, True)
elif cmd == 'put': put(args, link)
elif cmd == 'ls': ls(link, True)
elif cmd == 'find': find(args, link, True)
elif cmd == 'backup': backup(args, link)
elif cmd == 'flash': flash(args, link, False)
elif cmd == 'findget': findget(args, link, False)
elif cmd == 'findgetg': findget(args, link, True)
else: print('Unkown command: ' + cmd)



